Encryption/RIPEM3B4/ripem/doc/ripem.man



RIPEM(1)            UNIX Programmer's Manual             RIPEM(1)



NAME
     ripem - RSAREF-based Internet Privacy Enhanced Mail, version
     2.1.  Enciphers and deciphers messages using a public key
     encryption system, and formats them for embedding in elec-
     tronic mail messages.  Also generates RSA public keys.

SYNOPSIS
     ripem { -e | -d | -g | -G | -c }
          [ -b #_of_bits ] [ -v #_of_months ]
          [ -r recipient(s) ] [ -A { des-cbc | des-ede-cbc | rx2-cbc/bits } ]
          [ -m { encrypted | mic-only | mic-clear | enveloped-only} ]
          [ -M { ripem1 | pem | pkcs} ] [ -u myusername ] [ -h ipr ] [ -T amn ]
          [ -p publickey_infile ] [ -s privatekey_infile ]
          [ -H home_directory ]  [-B]
          [ -k {key_to_private_key|-} ] [ -K {new_key} ]
          [ -y pub_key_server_name ] [ -Y fgs ]
          [ -i infile ] [ -o outfile ]
          [ -D debug_level ] [ -Z debug_file ]
          [ -R cefkms ] [ -F random_file ] [ -C random_string ]
          [-x external_msg_file]  [-a dig_alg]
          <in >out

RIPEM Execution Mode Options
     -e   (encipher) means apply the transformations specified by
          the "-m" option to the input, writing to output an
          encapsulated message suitable for inclusion in an elec-
          tronic mail message.

     -d   (decipher) means process the input file, producing plaintext and
          inserting certificates and CRLs into the database as necessary.
          The input is expected to be an encapsulated PEM-style
          message.  If the "BEGIN PRIVACY-ENHANCED MESSAGE" boundary is
          not found, RIPEM will attempt to rewind and read the input as a
          PKCS message.  (This cannot be done if the input is stdin).
          You can go directly to reading as a PKCS message by using
          -M pkcs.

     -g   (generate) means generate an RSA key.  The private com-
          ponent will be written to the file privkey in the RIPEM
          home dir.  The public component will be written as a self-signed
          certificate to the file pubkeys in the RIPEM home dir.
          For the self-signed certificate, a "Persona Certificate"
          distinguished name is formed from your RIPEM user name.
          You should periodically back up all files in the RIPEM home dir.
          Note: key generation can be a time-consuming process.

     -G   (generate) means generate an RSA key just like -G, except RIPEM
          interactively lets you create a distinguished name of any form.

     -c   (change) means change the key to your private key.  The
          reencrypted private component will be written to the
          file privkey in the RIPEM home dir.  This option also re-
          displays the self-signed certificate digest.

Other RIPEM Options
     -A   (Algorithm) specifies the data encryption algorithm.
          Used only with "-e".  The legal values are "des-cbc",
          which is the default well-known DES in Cipher Block
          Chaining mode, "des-ede-cbc", "rx2-cbc/bits".  
          "des-ede-cbc" specifies Encrypt-Decrypt-Encrypt "Triple-DES" 
          as suggested by IBM a few years ago.  "rx2-cbc/bits"
          specifies RX2 which is an algorithm compatible with RC2
          from RSA Data Security.  The bits is the effective key
          size in bits.  For example, a typical value is rx2-cbc/40 .

          When "des-ede-cbc" is specified,
          "ripem1" and "pem" mode use DES-EDE2-CBC where each 64-bit block
          is enciphered with one 56-bit key, deciphered with a second
          56-bit key, and enciphered again with the first 56-bit key.
          Cipher Block Chaining is applied only after all three
          operations have been done.  The result is effectively a
          112-bit key, much more difficult to exhaustively search
          than single-DES.  Note that EDE-CBC mode is not widely used
          in PEM, and that there is even some disagreement as to just
          what crypto-graphic operations should be implied by this appella-
          tion.  Use EDE-CBC mode only when data security is much
          more important than compatibility, as EDE-CBC is not
          yet a PEM standard.

          When "des-ede-cbc" is specified,
          "pkcs" mode uses DES-EDE3-CBC where each 64-bit block
          is enciphered with one 56-bit key, deciphered with a second
          56-bit key, and enciphered again with a third 56-bit key.
          Cipher Block Chaining is applied only after all three
          operations have been done.  The result is effectively a
          168-bit key.  Support for DES-EDE3-CBC is explicitly recommended
          in the S/MIME implementor's guide.

     -a   (digest algorithm) specifies the digest algorithm used when
          creating a message (-e), creating new certificates or CRLS,
          or for verifying a PKCS detached signature
          "-d -M pkcs -x message-file" (see -x).  
          The legal values are "rsa-md5", which is the default MD5 message
          digest algorithm from RSA, "rsa-md2" which is the MD2 algorithm
          from RSA, and "sha-1" which is the SHA-1 algorithm.

     -B   (Binary message file) specifies that end-of-line character
          translation should not be performed on the message file.  Used
          only for PKCS messages (-M pkcs).  The behavior of the
          -B flag is undefined when using the standard input or output for
          the message data.

     -b   (bits) specifies the number of bits desired in the pub-
          lic key being created.  Used only with the "-g" option.
          Must be in the range 508 to 1024.  The default is a
          pseudorandom number of bits between 512 and 523.

     -C   (command line) specifies that the remainder of the com-
          mand line consists of pseudo-random strings to be used
          to construct message keys and so on.  This option must
          be the last option on the command line, as any argu-
          ments following it will be used only as pseudo-random
          data.

          The "-C" option should be used, when possible, with the
          command substitution capabilities of command shells
          such as Unix's C Shell.  The sequence "-C `ps -aux`
          `finger` `df`" would generate a reasonable amount of
          pseudo-random data on many Unix systems.

     -D   (debug) specifies the debug level.  0 disables debug
          output; larger values (up to about 4) specify increas-
          ing amounts of debug output.  The output is written to
          the destination specified by the "-Z" option.  The
          default level is 1, which gives you only messages
          regarding the key status of recipients and senders.

     -F   (file) specifies the name of a file containing pseudo-
          random information.  Used only for the "-e" and "-g"
          options.  A pseudo-random amount of data from this
          file, selected from pseudo-random locations in the
          file, will be used as a source of pseudo-random ini-
          tializing data for message keys and so on.  The file is
          used only if the "f" suboption of the "-R" option is
          specified.

          If -F is not given on the command line, RIPEM uses the
          file randomin in the RIPEM home directory. The random
          input file is checked for in this order: -F in
          RIPEM_ARGS, -F on the command line, randomin in the
          RIPEM home dir, and finally RIPEM_RANDOM_FILE environ-
          ment variable.

     -H   (home directory) specifies the RIPEM home directory
          where all the user's "local" data is kept, including
          the private key and certificates created by the user.
          The public and private keys files in the RIPEM home
          directory are searched before files given by -p and -s.

          If -H is not specified, the value of the environment
          variable RIPEM_HOME_DIR is used.  If this is not specified,
          ~/.ripemhome is used (on Unix) or C:\RIPEMHOM (on MS-DOS).

          If the RIPEM home directory does not exist, RIPEM
          creates it.

     -h   (header) specifies options for enciphering messages
          that contain mail headers.  Used only with the "-e"
          option.  Following -h must be a string of option
          letters selected from below:
          i
               (include) means include the message headers as
               part of the plaintext message--that is, enciper
               them.
          p
               (prepend) means place the unenciphered message
               headers at the beginning of the enciphered output,
               before the first Privacy Enhance Message Boundary.
               This option is useful when using RIPEM as a filter
               for messages that will be fed directly to a
               mailer.
          r
               (recipients) means scan the message headers for
               "To:" and "cc:" lines, and add these recipients to
               the recipient list.

     -i   (input) specifies input file.  Used by the "-e" and "-
          d" options.  Default is standard input.

     -K   (new key) specifies the new key used when changing the
          key to the private key.  Normally, RIPEM prompts the
          user interactively for the new key.  -K is intended to
          be used by wrapper programs that invoke RIPEM.  Use
          care with the -K and -k options, as you may inadver-
          tantly reveal the key to your private key to individu-
          als snooping on your computer.

     -k   (key) specifies the key used to encrypt the RSA private
          key that resides in the file referenced by the "-s"
          option.  If the "-k" option is not used, RIPEM looks
          for a non-empty environment variable named
          "RIPEM_KEY_TO_PRIVATE_KEY"; if it is found, its value
          is used as the key to decrypt the private component of
          the user's public key.  If neither the "-k" option nor
          the environment variable can be found, RIPEM prompts
          the user interactively for the private key.  Use of the
          "-k" option on a multiuser system is discouraged, as it
          increases the chances that another user can discover
          the key to your private key.

          As a special case, if the key is specified as "-",
          RIPEM will read the first line from standard input and
          use it as the key the the private key.  This capability
          is intended for use by programs that invoke RIPEM.

     -M   (message compatibility) specifies the message format of
          the outgoing message.  Used with the "-e" option.  (-M pkcs
          can also be used with the "-d" option to decipher as a PKCS
          message right away instead of trying PEM first.)

          "ripem1" specifies that the message is to be compatible
          with earlier versions of RIPEM.  This is the default.
          The Originator-Certificate field contains the sender's
          self-signed certificate.  This format will output
          Issuer-Certificate fields (which earlier versions of RIPEM
          ignore).

          "pem" specifies that the message is to be compatible
          with the RFC 1421 message syntax and the RFC 1422
          certification model.  Recipients are identified by
          Recipient-ID-Asymmetric fields, which give an issuer name
          and serial number (as opposed to "ripem1" which uses
          Recipient-Key-Asymmetric which gives the recipient's
          public key).  Since it is not known which issuer a
          recipient recognizes themself by, recipient fields are
          output for all possible issuers.  If the sender is certified
          under exactly one certificate chain, the Originator-
          Certificate field contains the certificate from the sender's
          immediate issuer.  However, if RIPEM finds certificates for
          the sender from more than one issuer, then it doesn't know
          which one to use, so it uses the self-signed certificate
          as in "ripem1".

          "pkcs" specifies that the message is to be compatible with
          the Public Key Cryptography Standards.  If -m is "encrypted",
          this creates a signed and enveloped message.  If -m is
          "mic-only" or "mic-clear", this creates a signed message.  If -m
          is "enveloped-only", this creates an enveloped-only message.

     -m   (mode) specifies the mode of the outgoing encrypted
          message.  Used only with the "-e" option.

          "encrypted" specifies that the message is to be
          encrypted using the algorithm given by -A, and that the message
          is to be "signed" computing a message digest of the
          input message, encrypting the digest with the sender's
          private key, and including the result in the outgoing message.
          For -M "ripem1" or "pem", the encrypted text of the message is
          encoded in RFC 1421 printable form and included in the output.
          (RFC 1421 printable form is similar to the well-known
          "uuencoding" format, but differs in a few details.)
          The MD5 message digest is always used upon output, but
          incoming messages using the MD2 algorithm can be pro-
          cessed correctly.  For -M "pkcs", this produces a PKCS
          signed and enveloped message using the same algorithms as above.

          "mic-only" specifies that the message is to be signed
          as described above, but not encrypted.  For -M "ripem1" or "pem"
          the text of the message is encoded in RFC 1421 form.  This option
          provides authentication but no confidentiality.   For -M "pkcs",
          this produces a PKCS signed message.

          "mic-clear" specifies that the message is to be signed as
          described above.  For -M "ripem1" or "pem", the text of the
          message is neither encrypted nor printably encoded, so the text
          of the message can be read by anyone with any standard mail
          program.  For -M "pkcs", this produces a PKCS signed message.

          "enveloped-only" is only valid for -M "pkcs" and specifies that
          the message is to be encrypted using the algorithm given by -A,
          but that the message is not signed.

          For all -m options, if -M "pkcs" is used, the PKCS message (which
          itself is binary output) is base64 encoded so that it
          can be mailed.

     -o   (output) specifies output file.  Used by the "-e" and
          "-d" options.  Default is standard output.

     -p   (public key) specifies a file containing the public
          components of users' RSA public keys to be searched after
          the file pubkeys in the RIPEM home dir has been searched.
          The program also looks in files given by the environment variable
          named RIPEM_PUBLIC_KEY_FILE if found.

          Multiple values can be specified; RIPEM will search
          these public key files in the order specified.

     -R   (random) specifies the source(s) of pseudo-random input
          used by RIPEM to generate message keys, initialization
          vectors, and RSA keys.  Used only with the "-e" and "-
          g" options.  Following -R must be a string of option
          letters selected from below:
          c
               (command) means use information from the command-
               line "-C" option.
          e
               (entire) means use information from the entire
               command line.
          f
               (file) means read information from a file; see the
               "-F" option.
          k
               (keyboard) means prompt the user for a string
               typed at the keyboard.
          m
               (message) means take characters from the input
               plaintext message as pseudo-random input.  (For -e
               only.)
          s
               (system) means query the system for system-
               dependent information.  The type and "randomness"
               of this information depends upon both the specific
               computer and the port of RIPEM that you are using.

          Default is "efms", but command line and file data are
          not used if the "-C" and "-F" arguments are not
          present.

     -r   (recipient) specifies the user to whom the message is
          being sent.  Used only with the "-e" option. The reci-
          pient name is usually specified as a complete email
          address.  The recipient name is used to lookup the
          recipient's public key component.  Multiple recipients
          can be specified, each with a separate "-r" option.

     -s   (secret key) specifies a file containing the user's secret
          (or private) component of the RSA keypair to be searched after
          the file privkey in the RIPEM home dir has been searched.
          The program also looks in files given by the environment variable
          named RIPEM_PRIVATE_KEY_FILE if found.

          Multiple values can be specified; RIPEM will search
          these private key files in the order specified.

     -T   (recipienT options) specifies options related to the
          processing of recipients.  Following -T must be a
          string of option letters selected from below:
          a
               (abort) specifies that RIPEM should uncondition-
               ally abort if, while in -e mode, it is unable to
               find the public key of one or more recipients.
               The default is to prompt the user for what to do.
          m
               (me) specifies that RIPEM should include the user
               as a recipient if encrypting in "-m encrypted"
               mode.  This is the default.
          n
               (none) specifies that RIPEM should clear the con-
               ditions that would be specified by "a" and "m"
               above.

     -u   (username) specifies your username--actually, your
          email address.  This is used to look up your self-signed
          certificate when logging into RIPEM.  For encipherment, the
          username is placed in the message header (along with
          the self-signed certificate) to inform the recipient of the
          sender's username.  For decipherment, the
          message header is searched for the username so that
          RIPEM will know which version of the message key to
          decrypt.

          If "-u" is not specified, RIPEM uses the value of the
          environment variable RIPEM_USER_NAME, if non-empty.
          Otherwise, the username defaults to
          <loginname>@<hostname> or "me" under MS-DOS and most
          other operating systems.

          The username can be specified as a comma-separated list
          of names.  In this case, the first name in the list is
          used as your username for -g and -e.  For -d, RIPEM

Encryption/RIPEM3B4/ripem/util/emacs/ripem.el
;; Functions for calling RIPEM from emacs mail.
;; Version 2.0
;; This file is public domain courtesy of RSA Data Security. Inc.

;;; Modified September 1994 by jefft@netcom.com
;;; Don't use -s and -p anymore.  The user can still set RIPEM_PUBLIC_KEY_FILE
;;;   and RIPEM_PRIVATE_KEY_FILE if they want to specify other files.
;;; Also, in ripem-receive, add one past the end of the end message boundary
;;;   to make sure it all gets fed to RIPEM.

;;; Modified Feb. 2 1994 by jefft@rsa.com
;;; Only use -s in change password if old private key file already exists.
;;; Remove generated-public-key-file from ripem-change-password

;;; Modified Dec. 1 1993 by jefft@rsa.com
;;; Added and use ripem-home-dir.
;;; Removed generated-public-key-file since this is now in ripem-home-dir.
;;; Added ripem-change-password and ripem-validate-and-receive.
;;; Always prompt for password when preparing or receiving.
;;; Modified ripem-list-users to include pubkeys in ripem-home-dir.
;;; Change ripem-process-region to allow displaying output in another
;;;   window.  This is used during receive so that the original encrypted
;;;   message is not erased.

;; Modified Mar  9 1993 by uri@watson.ibm.com
;;; Added normal interface to signing in clear:
;;;       ripem-sign-clear ()

;; Modified Jan 18 1993 by jefft@rsa.com  
;;; Fixed simple bug in ripem-list-users.

;; Modified Jan 6 1993 by Christopher Owens (owens@cs.uchicago.edu)
;;; 1) Replaced hard-coded file names with variables, and attempted to
;;;    do something vaguely reasonable with environment variables.
;;; 2) Eliminating the passing of the key-to-secret-key via the -k command
;;;    line option, which is a security hole.  Instead, we use the "-k -"
;;;    option, which tells RIPEM to read the key from standard input.
;;;    This requires some messing with a temp buffers, because we don't
;;;    want to (visibly) stuff the key into the current buffer, either.
;;; 3) Added progress indication
;;; 4) ripem-sign takes optional argument SIGN-CLEAR; if T then use
;;;    encryption mode mic-clear rather than mic-only
;;; 5) ripem-receive operates only on the body of the message (between the PEM
;;;    banners, and offers a bail-out option if those banners are not found
;;; 6) ripem-generate shows diagnostic output.

;;To install, add the expresion (load "/home/local/ripem/util/emacs/ripem.el") 
;;  to the file ~/.emacs or to /home/local/emacs/lisp/default, where
;;  /home/local/ripem/util/emacs is the directory where this file is, and
;;  /home/local/emacs is the directory where emacs is installed.

;; The ripem-executable is specified by ripem-program-name

;; The RIPEM home directory is specified by ripem-home-dir

;; Be sure to generate your keys first.  Use ESC-x ripem-generate.
;; The commands ESC-x ripem-sign and ripem-encrypt use the -h option when
;;   preparing the message, which means that the header with the To: fields,
;;   etc. should be at the top of the buffer, followed by a blank line or
;;   a line with "--text follows this line--".
;; Use ESC-x ripem-receive to decipher the RIPEM message in the buffer.
;; Use ESC-x ripem-list-users to display the user names in the public key file.
;; For more info, type C-h f and then the name of the function.

;;; Order of priority for finding file names for keys, etc:
;;; 1) If user has defined file name variables (i.e. in a .emacs file)
;;; 2) If user has defined environment variables
;;; 3) Defaults

(defvar ripem-home-dir
  (let ((name (getenv "RIPEM_HOME_DIR")))
    (if (null name)
        "~/.ripemhome"
      name))
  "Name of directory containing local RIPEM public and private key files")

(defvar ripem-program-name "ripem")
  
  
(defun ripem-sign (&optional clear)
  "Use this to replace the contents of your mail message buffer with a signed
  mail message, preserving the mail header.  Prefix arg means sign cleartext.
  If there is any error, use UNDO (i.e. type \\[undo])."
  (interactive "P")
  (open-after-mail-delimiter)
  (ripem-process-region (point-min)
                        (point-max)
                        t
                        "-e" "-m"
                        (if clear "mic-clear" "mic-only")
                        "-h" "p"))

(defun ripem-sign-clear ()
  "Use this to replace the contents of your mail message buffer with a signed
  mail message in cleartext (no encoding), preserving the mail header.
  If there is any error, use UNDO (i.e. type \\[undo])."
  (interactive)
  (open-after-mail-delimiter)
  (ripem-process-region (point-min)
                        (point-max)
                        t
                        "-e" "-m"
                        "mic-clear"
                        "-h" "p"))

(defun ripem-encrypt ()
  "Use this to replace the contents of your mail message buffer with an
encrypted mail message, preserving the mail header and encrypting for each
of the recipients in the To: and cc: fields.  These must be the full user
names as they appear in the public key file, such as alice@chirality.rsa.com.
(See ripem-list-users.)
  You must enter your private key password and a string of random bytes (for
generating a fresh encryption key).  You will see full name and validation
status of each of the recipients at the top of the buffer, which you must
examine for correctness.  When finished examining them, delete the names up
to the mail header using C-k and simply send the message as usual.
If there is any error, use UNDO (e.g. type \\[undo])"
  (interactive)
  (open-after-mail-delimiter)
  (let ((bytes (read-n-chars-no-echo
                "Enter %d random characters for encryption key" 16)))
    (ripem-process-region (point-min)
                          (point-max)
                          t
                          "-e"
                          "-m" "encrypted"
                          "-T" "am"
                          "-h" "pr"
                          "-R" "sc"
                          "-Y" "fsg"
                          "-C" bytes)
    (goto-char (point-min))
    ;; This inserts a message at the top of the buffer depending on whether
    ;;   RIPEM was successful.
    (if (looking-at "Recipient status:")
       (insert ">>>NOTE<<< Examine and remove recipient status info before sending. (use C-k)")
       (insert "(Use UNDO to return to previous contents, e.g. type C-_)"))
    (newline)
    (goto-char (point-min))))


(defun ripem-receive (&optional validation-mode)
  "Use this to decode the privacy enhanced message in the current buffer,
or to check signatures."
  (interactive)
  (save-excursion
    (let ((end (if (re-search-forward "END PRIVACY-ENHANCED MESSAGE-----$"
                                      (point-max) t)
                   (+ (match-end 0) 1)
                 nil))
          (begin (if (re-search-backward "^-----BEGIN PRIV" (point-min) t)
                     (max (point-min)
                          (- (match-beginning 0) 1))
                   nil)))
      (if (or (null begin) (null end))
          (if (y-or-n-p "Can't find headers; try to do whole buffer?")
              (setq begin (point-min)
                    end (point-max))
            (error "Couldn't find PEM headers in buffer")))
      (if validation-mode
	  (ripem-process-region begin end nil "-d" "-v" "24")
	(ripem-process-region begin end nil "-d")))))

(defun ripem-validate-and-receive ()
  "Use this after ripem-receive returns the message:
\"The following sender has not been validated: ...\"
  This will validate the sender of the privacy enhanced message in the
current buffer, adding them to the file public key file."
  (interactive)
  (ripem-receive t))



(defun ripem-process-region (start end replace-buffer-p &rest addl-args)
  "Copies the current region to a temp buffer, then prompt for the secret
   key password.
   Calls ripem, taking care of some housekeeping.
   If REPLACE-BUFFER-P is t, substitutes ripem's output for the current
   contents of region, else displays output in separate window."
  (let ((my-buffer (get-buffer-create " ripem-temp"))
        (oldbuf (current-buffer)))
    (save-excursion
      (set-buffer my-buffer)
      (erase-buffer)
      (buffer-flush-undo my-buffer)
      (insert (read-string-no-echo "Enter password to private key: "))
      (newline)
      (insert-buffer-substring oldbuf start end)
      (message "Running ripem ...")
      (apply 'call-process-region
             (point-min)
             (point-max)
             ripem-program-name
             t
             t
             nil
	     ;; Note: private key and local public keys should now be in
	     ;; the home dir.  See ripem-change-password
             "-H" (expand-file-name ripem-home-dir)
             "-k" "-"
             addl-args))
    (if (not replace-buffer-p)
	(display-buffer my-buffer)
      (goto-char start)
      (delete-region start end)
      (insert-buffer my-buffer)))
  (message "Running ripem ... done."))


(defun open-after-mail-delimiter ()
  "Ensure that there is an open line after the --text follows this line--
  delimiter in the mail buffer so that RIPEM will find the body."
  (goto-char (point-min))
  (and (re-search-forward "^--text follows this line--
." (point-max) t)
       ;; found a char on the beginning of the line which should be blank,
       ;;   so back up and insert a newline before it
       (goto-char (1- (point)))
       (insert "\n")))

(defun ripem-list-users ()
  "Show all the user names which can be used as recipients in ripem-encrypt or
  for verifying signatures in ripem-receive."
  (interactive)
  (shell-command (format "(cd %s; grep -h User: pubkeys)"
			 ripem-home-dir)))



(defun ripem-generate ()
  "Generates key files using the default user name.
You must enter your private key password and a long string of random bytes
  (for generating the keys).
Your username will have your full host name, such as alice@chirality.rsa.com.
  (If you want a different username, set up your RIPEM_USER_NAME environment
   variable and then generate.)
When done, archive copies of your private and public key files and
  don't forget your password!"
  (interactive)
  (let ((password (read-string-no-echo
		   "Enter password for private key (and don't forget it): " t))
	;; get 512 bits of random data, assuming 4 usable bits per char
	(random-chars (read-n-chars-no-echo
		       "Enter %d random characters for generating keys" 128))
        (buffer (get-buffer-create " ripem-temp")))
    ;; Create ripem-home-dir.  Ignore message if it already exists.
    (shell-command (format "mkdir -p %s > /dev/null" ripem-home-dir))
    (save-excursion
      (set-buffer buffer)
      (erase-buffer)
      (buffer-flush-undo buffer)
      (insert password)
      (newline)
      (message "Generating ...")
      (call-process-region (point-min) (point-max) ripem-program-name
                           t t t
                           "-g"
                           "-H" (expand-file-name ripem-home-dir)
                           "-k" "-"
                           "-R" "sc"
                           "-C" random-chars)
      (newline)
      (insert (format "*** You must now archive copies of pubkeys and privkey in %s"
                       ripem-home-dir))
      (newline)
      (insert "Record the certificate digest to give to others who want to validate your key.")
      (newline)
      (insert "If you forget your certificate digest, do ripem-change-password to see it.")
      (newline))
    (display-buffer buffer)
    (message "Generating. .. done.")))

(defun ripem-change-password ()
  "Change the password to the user's private key.
You must enter your old and new private key passwords.
This displays your certificate digest (and will create a new self-signed
  certificate if one doesn't exist).
When done, archive copies of your new private and public key files and
  don't forget your password!"
  (interactive)
  (let ((password (read-string-no-echo "Enter password for private key: "))
	(new-password (read-string-no-echo
	       "Enter new password for private key (and don't forget it): " t))
        (buffer (get-buffer-create " ripem-temp")))
    ;; Create ripem-home-dir.  Ignore message if it already exists.
    (shell-command (format "mkdir -p %s > /dev/null" ripem-home-dir))
    (save-excursion
      (set-buffer buffer)
      (erase-buffer)
      (buffer-flush-undo buffer)
      (insert password)
      (newline)
      (call-process-region
	     (point-min) (point-max) ripem-program-name
	     t t t
	     "-c"
	     "-H" (expand-file-name ripem-home-dir)
	     "-k" "-"
	     "-K" new-password)
      (newline)
      (insert (format "*** You must now archive copies of pubkeys and privkey in %s"
                       ripem-home-dir))
      (newline)
      (insert "The certificate digest can be given to others who want to validate your key.")
      (newline))
    (display-buffer buffer)
    (message "Done.")))

(defun read-string-no-echo (prompt &optional confirm)
  "Read a string from the minibuffer, prompting with PROMPT.
Optional second argument CONFIRM non-nil means that the user will be asked
  to type the string a second time for confirmation and if there is a
  mismatch, the process is repeated.

Line editing keys are:
  C-h, DEL      rubout
  C-u, C-x      line kill
  C-q, C-v      literal next"
  (catch 'return-value
    (save-excursion
      (let ((input-buffer (get-buffer-create " *password*"))
            (cursor-in-echo-area t)
            (echo-keystrokes 0)
            char string help-form done kill-ring)
        (set-buffer input-buffer)
        (unwind-protect
            (while t
              (erase-buffer)
              (message prompt)
              (while (not (memq (setq char (read-char)) '(?\C-m ?\C-j)))
                (if (setq form
                          (cdr
                           (assq char
                                 '((?\C-h . (delete-char -1))
                                   (?\C-? . (delete-char -1))
                                   (?\C-u . (delete-region 1 (point)))
                                   (?\C-x . (delete-region 1 (point)))
                                   (?\C-q . (quoted-insert 1))
                                   (?\C-v . (quoted-insert 1))))))
                    (condition-case error-data
                        (eval form)
                      (error t))
                  (insert char))
                (message prompt))
              (cond ((and confirm string)
                     (cond ((not (string= string (buffer-string)))
                            (message
                             (concat prompt "[Mismatch... try again.]"))
                            (ding)
                            (sit-for 2)
                            (setq string nil))
			   ((string= string "")
                            (message
                             (concat prompt
				     "[String cannot be empty... try again.]"))
                            (ding)
                            (sit-for 2)
                            (setq string nil))
                           (t (throw 'return-value string))))
                    (confirm
                     (setq string (buffer-string))
                     (message (concat prompt "[Retype to confirm...]"))
                     (sit-for 2))
                    (t (throw 'return-value (buffer-string)))))
          (message (concat prompt "[Thank you!]"))
          (set-buffer-modified-p nil)
          (kill-buffer input-buffer))))))

(defun read-n-chars-no-echo (prompt n)
  "Display prompt and return a string of n characters, without echoing.
If prompt has %d in it, the countdown from n is printed there.
After n chars have been read, this still adds them to the string, and
  this waits for the user to hit enter."
  (let ((echo-keystrokes 0)
	(string "")
	(count n)
	char)
    (while (> count 0)
      (message prompt count)
      (setq string (concat string (list (read-char))))
      (setq count (1- count)))
    (while (progn
	     (message "Hit ENTER when done")
	     (not (or (= (setq char (read-char)) 13) (= char 10))))
      (setq string (concat string (list char))))
    (message "Hit ENTER when done [Thank you!]")    
    string))
